# Digital House - Trabajo Práctico Nº1
## Desafío 1. Análisis exploratorio de un dataset de precios de propiedades
Grupo Nª2
Alumnos:
* Alberto Paparelli
* André Song
* Marianela Riz
* Gloria Caravajal
* Franco Emanuel Rodriguez
* Iván Axel Schweizer

Objetivos:
* Efectuar una limpieza del dataset provisto. Particularmente, deberá diseñar estrategias para lidiar con los datos perdidos en ciertas variables.
* Realizar un análisis descriptivo de las principales variables.
* Crear nuevas columnas a partir de las características dadas que puedan tener valor
predictivo.

In [39]:
# Lo primero que hacemos es importar todas las librerias necesarias.
import pandas as pd

In [40]:
# Abrimos el dataset y hacemos un chequeo rápido para saber cantidad de registros y miramos la composición del mismo.
filename = "../dataset/properati.csv"
data = pd.read_csv(filename)
display(data.sample(n=6)) 
print(f"(Cantidad de Filas, Cantidad de columnas) -> {data.shape}")


Unnamed: 0 operation property_type       place_name  \
117899      117899      sell         house     San Fernando   
69471        69471      sell         house  Lomas de Zamora   
114285      114285      sell     apartment         Belgrano   
61804        61804      sell         house          Miramar   
9984          9984      sell     apartment        Caballito   
13746        13746      sell     apartment         Recoleta   

                                  place_with_parent_names country_name  \
117899  |Argentina|Bs.As. G.B.A. Zona Norte|San Fernan...    Argentina   
69471   |Argentina|Bs.As. G.B.A. Zona Sur|Lomas de Zam...    Argentina   
114285               |Argentina|Capital Federal|Belgrano|    Argentina   
61804    |Argentina|Buenos Aires Costa Atlántica|Miramar|    Argentina   
9984                |Argentina|Capital Federal|Caballito|    Argentina   
13746                |Argentina|Capital Federal|Recoleta|    Argentina   

                          state_name  geonames_id  \
117899      Bs.As. G.B.A. Zona Norte          NaN   
69471         Bs.As. G.B.A. Zona Sur          NaN   
114285               Capital Federal    3436077.0   
61804   Buenos Aires Costa Atlántica    3430668.0   
9984                 Capital Federal    3435874.0   
13746                Capital Federal    3429595.0   

                              lat-lon        lat  ...  surface_covered_in_m2  \
117899                            NaN        NaN  ...                  123.0   
69471                             NaN        NaN  ...                    NaN   
114285                            NaN        NaN  ...                  341.0   
61804         -38.2783484,-57.8381089 -38.278348  ...                    NaN   
9984    -34.6150650909,-58.4603789864 -34.615065  ...                   35.0   
13746         -34.5906397,-58.3785474 -34.590640  ...                  280.0   

        price_usd_per_m2 price_per_m2  floor  rooms  expenses  \
117899       1423.076923  1504.065041    NaN    NaN    3700.0   
69471                NaN          NaN    NaN    NaN       NaN   
114285               NaN  1554.252199    2.0    4.0       NaN   
61804                NaN          NaN    NaN    3.0       NaN   
9984         2265.909091  2848.571429    NaN    NaN       NaN   
13746        3000.000000  3214.285714    NaN    6.0       NaN   

                                            properati_url  \
117899  http://www.properati.com.ar/1ce4k_venta_casa_s...   
69471   http://www.properati.com.ar/1aiag_venta_casa_l...   
114285  http://www.properati.com.ar/1c9oq_venta_depart...   
61804   http://www.properati.com.ar/1a65i_venta_casa_m...   
9984    http://www.properati.com.ar/1663q_venta_depart...   
13746   http://www.properati.com.ar/16q95_venta_depart...   

                                              description  \
117899  Excelente oportunidad! Triplex de 3 Dormitorio...   
69471   Casa en venta de 3 ambientes sobre la calle Lo...   
114285  Corredor Responsable: Gustavo Guastello - C.U....   
61804   Propiedad a la venta Ubicada en calle 33 numer...   
9984    Departamento monoambiente al frente con salida...   
13746   EXCLUSIVO PISO EN TORRRE DE LUJOLa mejor de to...   

                                                    title  \
117899  RUCA INMUEBLES | Excelente Casa 3 Dom c/ Coche...   
69471                                       Loria N° 2160   
114285                Departamento 4 amb c/dep, y cochera   
61804                                 Propiedad en venta    
9984    Trelles, Manuel R. al 1000, entre Av. Gaona y ...   
13746   BARRIO NORTE - VENTA  DEPARTAMENTO PISO ALTO  ...   

                                          image_thumbnail  
117899  https://thumbs4.properati.com/7/b_HxPyE9KvjHUv...  
69471   https://thumbs4.properati.com/7/1DXVkFlQuuQARq...  
114285  https://thumbs4.properati.com/7/_SN8vfUtL6pcCw...  
61804   https://thumbs4.properati.com/4/gfe9bwq_C3kgyw...  
9984    https://thumbs4.properati.com/1/r2j0PkCdpB4szp...  
13746   https:

(Cantidad de Filas, Cantidad de columnas) -> (121220, 26)


In [66]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121220 entries, 0 to 121219
Data columns (total 26 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Unnamed: 0                  121220 non-null  int64  
 1   operation                   121220 non-null  object 
 2   property_type               121220 non-null  object 
 3   place_name                  121197 non-null  object 
 4   place_with_parent_names     121220 non-null  object 
 5   country_name                121220 non-null  object 
 6   state_name                  121220 non-null  object 
 7   geonames_id                 102503 non-null  float64
 8   lat-lon                     69670 non-null   object 
 9   lat                         69670 non-null   float64
 10  lon                         69670 non-null   float64
 11  price                       100810 non-null  float64
 12  currency                    100809 non-null  object 
 13  price_aprox_lo

In [70]:
data.describe()

Unnamed: 0   geonames_id           lat           lon         price  \
count  121220.000000  1.025030e+05  69670.000000  69670.000000  1.008100e+05   
mean    60609.500000  3.574442e+06    -34.626210    -59.266290  4.685259e+05   
std     34993.344153  3.541306e+05      1.980936      2.299922  2.260101e+06   
min         0.000000  3.427208e+06    -54.823985    -75.678931  0.000000e+00   
25%     30304.750000  3.430234e+06    -34.669065    -58.727040  1.100000e+05   
50%     60609.500000  3.433910e+06    -34.597985    -58.480128  1.850000e+05   
75%     90914.250000  3.836668e+06    -34.441299    -58.395908  4.200000e+05   
max    121219.000000  6.948895e+06      4.545843    -53.733330  6.500000e+08   

       price_aprox_local_currency  price_aprox_usd  surface_total_in_m2  \
count                1.008100e+05     1.008100e+05         81892.000000   
mean                 4.229397e+06     2.397006e+05           233.795328   
std                  6.904714e+06     3.913239e+05          1782.222147   
min                  0.000000e+00     0.000000e+00             0.000000   
25%                  1.583309e+06     8.973388e+04            50.000000   
50%                  2.558452e+06     1.450000e+05            84.000000   
75%                  4.675792e+06     2.650000e+05           200.000000   
max                  8.212711e+08     4.654544e+07        200000.000000   

       surface_covered_in_m2  price_usd_per_m2  price_per_m2        floor  \
count          101313.000000      68617.000000  8.765800e+04  7899.000000   
mean              133.050181       2160.086916  6.912216e+03    17.452336   
std               724.351479       2759.288621  2.837864e+04   120.243621   
min                 0.000000          0.600000  1.510204e+00     1.000000   
25%                45.000000       1218.181818  1.550000e+03     1.000000   
50%                75.000000       1800.000000  2.213115e+03     3.000000   
75%               150.000000       2486.411765  3.355549e+03     6.000000   
max            187000.000000     206333.333333  4.000000e+06  3150.000000   

              rooms      expenses  
count  47390.000000  1.426200e+04  
mean       3.080840  5.009234e+03  
std        1.860773  1.204403e+05  
min        1.000000  1.000000e+00  
25%        2.000000  1.000000e+03  
50%        3.000000  2.000000e+03  
75%        4.000000  4.000000e+03  
max       32.000000  1.000150e+07

In [ ]:
data.isnull().mean()*100

Unnamed: 0                     0.000000
operation                      0.000000
property_type                  0.000000
place_name                     0.018974
place_with_parent_names        0.000000
country_name                   0.000000
state_name                     0.000000
geonames_id                   15.440521
lat-lon                       42.525986
lat                           42.525986
lon                           42.525986
price                         16.837156
currency                      16.837981
price_aprox_local_currency    16.837156
price_aprox_usd               16.837156
surface_total_in_m2           32.443491
surface_covered_in_m2         16.422208
price_usd_per_m2              43.394654
price_per_m2                  27.686850
floor                         93.483749
rooms                         60.905791
expenses                      88.234615
properati_url                  0.000000
description                    0.001650
title                          0.000000


In [69]:
data.describe()

Unnamed: 0   geonames_id           lat           lon         price  \
count  121220.000000  1.025030e+05  69670.000000  69670.000000  1.008100e+05   
mean    60609.500000  3.574442e+06    -34.626210    -59.266290  4.685259e+05   
std     34993.344153  3.541306e+05      1.980936      2.299922  2.260101e+06   
min         0.000000  3.427208e+06    -54.823985    -75.678931  0.000000e+00   
25%     30304.750000  3.430234e+06    -34.669065    -58.727040  1.100000e+05   
50%     60609.500000  3.433910e+06    -34.597985    -58.480128  1.850000e+05   
75%     90914.250000  3.836668e+06    -34.441299    -58.395908  4.200000e+05   
max    121219.000000  6.948895e+06      4.545843    -53.733330  6.500000e+08   

       price_aprox_local_currency  price_aprox_usd  surface_total_in_m2  \
count                1.008100e+05     1.008100e+05         81892.000000   
mean                 4.229397e+06     2.397006e+05           233.795328   
std                  6.904714e+06     3.913239e+05          1782.222147   
min                  0.000000e+00     0.000000e+00             0.000000   
25%                  1.583309e+06     8.973388e+04            50.000000   
50%                  2.558452e+06     1.450000e+05            84.000000   
75%                  4.675792e+06     2.650000e+05           200.000000   
max                  8.212711e+08     4.654544e+07        200000.000000   

       surface_covered_in_m2  price_usd_per_m2  price_per_m2        floor  \
count          101313.000000      68617.000000  8.765800e+04  7899.000000   
mean              133.050181       2160.086916  6.912216e+03    17.452336   
std               724.351479       2759.288621  2.837864e+04   120.243621   
min                 0.000000          0.600000  1.510204e+00     1.000000   
25%                45.000000       1218.181818  1.550000e+03     1.000000   
50%                75.000000       1800.000000  2.213115e+03     3.000000   
75%               150.000000       2486.411765  3.355549e+03     6.000000   
max            187000.000000     206333.333333  4.000000e+06  3150.000000   

              rooms      expenses  
count  47390.000000  1.426200e+04  
mean       3.080840  5.009234e+03  
std        1.860773  1.204403e+05  
min        1.000000  1.000000e+00  
25%        2.000000  1.000000e+03  
50%        3.000000  2.000000e+03  
75%        4.000000  4.000000e+03  
max       32.000000  1.000150e+07

MARU:

In [72]:
#1ero: Intentar rellenar las columnas lo mas que se pueda con info del mismo data o internet
#2do: eliminar posibles duplicados
#3ero: mandar columnas que queremos eliminar a futuro a una nueva lista

#COLUMNAS MARU:
#Rellenar geonames_id con lat, lon, dsp eliminamos estas de lat y lon y me quedo con la grupal que es lat-lon.
#Rellenar Floor y Room 

# Tipo de info de cada columna 


In [74]:
data['geonames_id'].value_counts()

3838574.0    8504
3860259.0    6606
3430863.0    6534
3430234.0    3279
3436077.0    2992
             ... 
3834971.0       1
3835354.0       1
3835711.0       1
3855135.0       1
3430940.0       1
Name: geonames_id, Length: 646, dtype: int64

In [75]:
# Columna operation: revisamos que opciones tiene, o si se repite lo mismo.
print(f'Cantidad de registros con NaN: {data.geonames_id.isnull().sum()}')
print("Contamos cantidad de registros unicos:")
print(data.groupby(['geonames_id']).size().reset_index(name='count'))


Cantidad de registros con NaN: 18717
Contamos cantidad de registros unicos:
     geonames_id  count
0      3427208.0      7
1      3427212.0      8
2      3427326.0    139
3      3427354.0     41
4      3427366.0     15
..           ...    ...
641    6693230.0     57
642    6942538.0     17
643    6942539.0     15
644    6942842.0      2
645    6948895.0      2

[646 rows x 2 columns]


LAT-LON:

In [49]:
data['lat-lon'].value_counts()

-34.4026444,-58.6684776          312
-34.40014,-58.64985              271
-34.6428675967,-58.4376599743    259
-34.425087,-58.5796585           224
-34.5895459,-58.3973636          156
                                ... 
-34.5933338,-58.4430338            1
-34.5513449,-58.6876066            1
-38.0801129219,-57.5860380405      1
-38.020255,-57.5281042             1
-34.6295045177,-58.460641832       1
Name: lat-lon, Length: 47203, dtype: int64

In [50]:
url_page= "https://nominatim.openstreetmap.org/search.php?q=-31.412513,-64.171582&polygon_geojson=1&format=json"

LAT:

In [51]:
data['lat'].value_counts()

-34.402644    312
-34.400140    271
-34.642868    259
-34.425087    224
-34.589546    157
             ... 
-34.536701      1
-34.647220      1
-34.880130      1
-34.678710      1
-34.629505      1
Name: lat, Length: 46495, dtype: int64

In [60]:
# Columna operation: revisamos que opciones tiene, o si se repite lo mismo.
print(f'Cantidad de registros con NaN: {data.lat.isnull().sum()}')
print("Contamos cantidad de registros unicos:")
print(data.groupby(['lat']).size().reset_index(name='count'))


Cantidad de registros con NaN: 51550
Contamos cantidad de registros unicos:
             lat  count
0     -54.823985      1
1     -54.819736      2
2     -54.806598      1
3     -54.806017      1
4     -54.801452      1
...          ...    ...
46490 -24.168071      1
46491 -24.166966      1
46492 -24.165218      1
46493 -23.137164      2
46494   4.545843      1

[46495 rows x 2 columns]


LON:

In [52]:
data['lon'].value_counts()

-58.668478    312
-58.649850    272
-58.437660    259
-58.579659    225
-58.397364    176
             ... 
-58.377785      1
-58.488385      1
-58.694755      1
-58.451329      1
-58.460642      1
Name: lon, Length: 46522, dtype: int64

In [61]:
# Columna operation: revisamos que opciones tiene, o si se repite lo mismo.
print(f'Cantidad de registros con NaN: {data.lon.isnull().sum()}')
print("Contamos cantidad de registros unicos:")
print(data.groupby(['lon']).size().reset_index(name='count'))

Cantidad de registros con NaN: 51550
Contamos cantidad de registros unicos:
             lon  count
0     -75.678931      1
1     -71.665143      1
2     -71.664029      1
3     -71.663814      1
4     -71.663515     15
...          ...    ...
46517 -54.580379      1
46518 -54.560043      1
46519 -54.559587      1
46520 -54.556127      1
46521 -53.733330      1

[46522 rows x 2 columns]


# TODO 
completar análisis de columnas: que columnas hay y que significan cada una.


Analizamos contenido de las columnas que creemos que pueden eliminarse.

Detalle de análisis de columnas:
* `operation` (sell/ventas) -> Vemos que no tiene NaNs y solo tiene la opción `sell`. Por este motivo la vamos a eliminar.

* `country_name` (nombre de paises) -> Vemos que no tiene NaNs y solo tiene la opción `Argentina`. Por este motivo la podriamos eliminar.

* `state_name` (nombre de estados) -> Vemos que no tiene NaNs.

* `place_name` (localidades) -> Vemos que tiene 23 NaNs.

* `place_with_parent_names` (nombre completo) -> Vemos que no tiene NaNs. El mismo esta formado con las columnas 
country_name|state_name|place_name. Podriamos utilizar esta columna eliminando o desestimando las 3 anteriores. 

* `geonames` (geonombres) -> Vemos que tiene 18717 NaNs. 

* `lat` (latitud) -> Vemos que tiene 51550 NaNs. Si comparamos los valores de la columna lon con los de la columna lat-lon, podemos observar que difieren 18025 valores. Los mismos presentan diferencias por "," en vez de ".", esto podria ser ocasionado por un mal tipeo en la carga de datos. 

* `lon` (longitud) -> Vemos que tiene 51550 NaNs. Si comparamos los valores de la columna lon con los de la columna lat-lon, podemos observar que difieren 27485 valores. Los mismos presentan diferencias por "," en vez de ".", esto podria ser ocasionado por un mal tipeo en la carga de datos. 

* `lat-lon` (latitud-longitud) -> Podemos confirmar que por algun error estos valores no coinciden con los desplegados en la columna lat y lon, por ende debemos evaluar que hacemos con estas 3 columnas, con que valor certero nos quedamos, con lat y lon o con lat-lon?

* `surface_total_in_m2` (superficie total m2) --> Vemos que tiene 19907 NaNs.

* `surface_covered_in_m2` (superficie m2) --> Vemos que tiene 19907 NaNs.




In [ ]:
mask_lat_lon=data.apply(lambda x: x["lat-lon"]== x["lat"] + "-" + x["lon"] if any(x["lat-lon"],x["lat"], x["lon"])==np.NaN)

In [ ]:
mask=data.apply(lambda x: x["lat-lon"]== x["lat"] + "-" + x["lon"])
mask